<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: **SUPER IMPORTANTE - notebooks sin nombre no serán revisados**

- Nombre de alumno 1: Diego Bartolucci
- Nombre de alumno 2: Pilar Nilo


### **Link de repositorio de GitHub:** [Repositorio](https://github.com/DiegoBarto01/MDS7202-pili-barto.git)

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Jueves a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
%pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.0/19.0 MB 28.6 MB/s eta 0:00:00
  Attempting uninstall: plotly
    Found existing installation: plotly 5.15.0
    Uninstalling plotly-5.15.0:
      Successfully uninstalled plotly-5.15.0


# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [2]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = '/content/drive/MyDrive/Lab 3 - Pandas/'
except:
    print('Ignorando conexión drive-colab')

Mounted at /content/drive


In [ ]:
# Si usted está en su máquina local, puede modificar la variable path a la carpeta donde tiene los archivos del lab.
#path = './'

## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://i.imgur.com/44knnIf.png" width="500" height="500">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [3]:
df_grades = pd.read_json(path + "students_grades.json")


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **1000** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **1000** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier registro que posea un valor diferente a esta escala, **debería ser eliminado de la base de datos**.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [4]:
def exploratory_data_analysis(dataframe):
  #1. dimension df
  print(f"1. La dimension del dataframe es: {dataframe.shape} \n")
  #2. nombres de las columnas del df
  print(f"2. Las columnas que conforman el dataframe son: \n {dataframe.columns} \n")
  #3. mostrar los primeros 5 elementos, 5 ultimos elementos y 5 elementos aleatoreos
  print("3.a Los 5 primeros elementos del dataframe son:")
  display(dataframe.head(5))
  print("3.b Los 5 últimos elementos del dataframe son:")
  display(dataframe.tail(5))
  print("3.c 5 elementos al azar del dataframe son:")
  display(dataframe.sample(5, random_state=10))
  #4. descripcion rapida de las variables numericas del dataframe
  print(f"4. Descripción de las variables numéricas del dataframe:")
  display(dataframe.describe())
  #5. mensaje con la cantidad de valores nulos
  print(f"5. La cantidad de valores nulos en el dataframe son: \n {dataframe.isna().sum()} \n")
  #6. mensaje que detalle la cantidad de valores unicos
  print(f"6. La cantidad de valores unicos en el dataframe son: \n {dataframe.nunique()} \n")
  #7. mensaje con el total de filas que contengan nombres duplicados
  print(f"7. Total de filas que contienen nombres duplicados: {dataframe.duplicated(subset=['names']).sum()} \n")
  #8. Mostrar las filas que contengan nombres sean duplicados
  print(f"8. Filas que contienen nombres duplicados:")
  display(dataframe[dataframe.duplicated(subset=['names'])])


In [5]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

In [6]:
exploratory_data_analysis(df_grades)

1. La dimension del dataframe es: (1400, 9) 

2. Las columnas que conforman el dataframe son: 
 Index(['names', 'gender', 'race/ethnicity', 'parental level of education',
       'lunch', 'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object') 

3.a Los 5 primeros elementos del dataframe son:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166

3.b Los 5 últimos elementos del dataframe son:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76

3.c 5 elementos al azar del dataframe son:


names  gender race/ethnicity parental level of education  \
7         Wayne Wilson    male        group B            some high school   
1325   Nicholas Reeves    male        group C                 high school   
784         Paul Smith    male        group B                 high school   
307       David Osteen    male        group C                 high school   
684   Marlene Sullivan  female        group D             master's degree   

         lunch test preparation course math score reading score writing score  
7     standard               completed         94            86            87  
1325  standard                    none        124           142           193  
784   standard                    none         70            65            60  
307   standard                    none         71            66            65  
684   standard               completed         77            82            91

4. Descripción de las variables numéricas del dataframe:


names  gender race/ethnicity parental level of education  \
count             1400    1400           1400                        1400   
unique            1153       2              5                           6   
top     Robert Frazier  female        group C                some college   
freq                 3     723            452                         334   

           lunch test preparation course  math score  reading score  \
count       1400                    1400        1400           1400   
unique         2                       2          97             89   
top     standard                    none         193            193   
freq         904                     898         120            117   

        writing score  
count            1400  
unique             93  
top               193  
freq              133

5. La cantidad de valores nulos en el dataframe son: 
 names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64 

6. La cantidad de valores unicos en el dataframe son: 
 names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       97
reading score                    89
writing score                    93
dtype: int64 

7. Total de filas que contienen nombres duplicados: 247 

8. Filas que contienen nombres duplicados:


names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course math score reading score  \
12        standard                    none        193           193   
81        standard                    none         82            85   
160       standard               completed         60            70   
170   free/reduced                    none         43            60   
176       standard                    none         54            61   
...            ...                     ...        ...           ...   
1385      standard                    none        193           147   
1390      standard                    none        124            64   
1393      standard               completed        124           193   
1395      standard                    none        193           142   
1398  free/reduced                    none        193           196   

     writing score  
12             193  
81              87  
160             74  
170             58  
176             58  
...            ...  
1385           131  
1390           193  
1393            49  
1395           193  
1398           193  

[247 rows x 9 columns]

In [ ]:
# df_grades[df_grades["names"]=="Elizabeth Quintero"]

**Justificación:**

```
 9. Las anomalías que se observan en los datos del dataframe "df_grades" es que existen 1400 registros cuando hay 1000 matrículas y de los 1400 valores solo 1153 son unicos, por lo cual existen valores duplicados. Además por lo que se observa, no existen valores nulos y de los valores duplicados que se muestran, existen notas mayores a 100, lo cual no puede suceder.

 10. Lo que se puede interpretar del gráfico de la distribución de las notas es que desde el primer y tercer cuartil están entre los 50 y 100 puntos, para más detalle la línea de partición de cada caja, que indica la mediana está entre los 70 puntos aproximadamente. Además, nos muestra puntajes minimos, máximos y los que se encuentran por sobre el 75% y bajo el 25% del total de alumnos.
 Con lo anterior, se puede visualizar perfectamente que hay puntajes que no corresponden, como por ejemplo los puntajes sobre 100pts. en cada tipo de prueba.
```

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y **cree una función** que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**
- [ ] Eliminar registros con valores nulos.
- [ ] Eliminar valores fuera del rango válido.
- [ ] Eliminar registros con nombres duplicados
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hints**

- Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.
- Cuidado con el orden en la limpieza de datos. ¿Que se debe eliminar primero, los nombres duplicados o los registros fuera del rango posible?


> Utilice la función creada en el punto 1 para verificar sus resultados.

**Respuesta:**

In [7]:
# Escriba su código aquí
def clean_data(dataframe):
    #1. Eliminar registros con valores nulos
    dataframe=dataframe.replace(" ", np.nan)
    dataframe=dataframe.dropna()
    #2. Eliminar valores fuera del rango válido
        #columnas que necesito verificar puntaje
    columnas_interes=['math score', 'reading score', 'writing score']
        #los paso a formato numero(float)
    dataframe[columnas_interes]=dataframe[columnas_interes].apply(pd.to_numeric, errors='coerce')
        #creo una mascara
    boolean_mask=(dataframe.loc[ :,"math score"]<=100 ) & (dataframe.loc[:, "reading score"] <=100) & (dataframe.loc[:, "writing score"] <=100)
        #lo aplico al dataframe
    dataframe=dataframe[boolean_mask].astype(object)

    #3. Eliminar registros con nombres duplicados
    return dataframe[~dataframe.duplicated(subset=['names'])]


In [9]:
#Pruebe su código aquí
df_grades = clean_data(df_grades)
#df_grades.dtypes

In [10]:
#4. Comprobar que no existen anomalias en los datos a traves de la funcion de la seccion 1.1
exploratory_data_analysis(df_grades)

1. La dimension del dataframe es: (1000, 9) 

2. Las columnas que conforman el dataframe son: 
 Index(['names', 'gender', 'race/ethnicity', 'parental level of education',
       'lunch', 'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object') 

3.a Los 5 primeros elementos del dataframe son:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course math score reading score writing score  
0  standard                    none       37.0          46.0          46.0  
3  standard               completed       80.0          78.0          81.0  
5  standard                    none       73.0          66.0          63.0  
6  standard               completed       81.0          75.0          76.0  
7  standard               completed       94.0          86.0          87.0

3.b Los 5 últimos elementos del dataframe son:


names  gender race/ethnicity parental level of education  \
1392   Richard Young    male        group D                 high school   
1394   Wanda Russell  female        group B                 high school   
1396  Marina Zeigler  female        group C           bachelor's degree   
1397   Laurie Carter  female        group B            some high school   
1399    Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
1392      standard                    none       69.0          75.0   
1394  free/reduced               completed       23.0          44.0   
1396  free/reduced               completed       66.0          74.0   
1397      standard               completed       54.0          61.0   
1399      standard               completed       68.0          80.0   

     writing score  
1392          71.0  
1394          36.0  
1396          81.0  
1397          62.0  
1399          76.0

3.c 5 elementos al azar del dataframe son:


names  gender race/ethnicity parental level of education  \
1186    George Lindsay    male        group E            some high school   
1347    Judith Mcenany  female        group D           bachelor's degree   
780   Richard Anderson    male        group C            some high school   
250      Maxine Walton  female        group E            some high school   
1075     Loretta Velez  female        group C          associate's degree   

             lunch test preparation course math score reading score  \
1186  free/reduced               completed       73.0          67.0   
1347      standard                    none       65.0          67.0   
780       standard               completed       67.0          73.0   
250   free/reduced                    none       74.0          74.0   
1075      standard               completed       96.0          96.0   

     writing score  
1186          59.0  
1347          62.0  
780           68.0  
250           72.0  
1075          99.0

4. Descripción de las variables numéricas del dataframe:


names  gender race/ethnicity parental level of education  \
count            1000    1000           1000                        1000   
unique           1000       2              5                           6   
top     Rita Courtney  female        group C                some college   
freq                1     518            319                         226   

           lunch test preparation course  math score  reading score  \
count       1000                    1000      1000.0         1000.0   
unique         2                       2        81.0           72.0   
top     standard                    none        65.0           72.0   
freq         645                     642        36.0           34.0   

        writing score  
count          1000.0  
unique           77.0  
top              74.0  
freq             35.0

5. La cantidad de valores nulos en el dataframe son: 
 names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64 

6. La cantidad de valores unicos en el dataframe son: 
 names                          1000
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       81
reading score                    72
writing score                    77
dtype: int64 

7. Total de filas que contienen nombres duplicados: 0 

8. Filas que contienen nombres duplicados:


Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

**Justificación:**

```
Primero se deben eliminar los registros que tienen puntajes fuera del rango ya que si borro los nombres duplicados primero, se corre el riesgo de que se elimine un puntaje valido (dentro del rango) y que quede un valor fuera del rango por ende a la hora de aplicar la eliminación de los puntajes se eliminará el registro de una matricula. Con lo anterior habrán menos de 1000 matriculas por ende se perderían datos, por lo tanto es mejor eliminar primero los registros fuera de rango y luego los duplicados.
```

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [11]:
# Escriba su código aquí
def nota_chilena(dataframe_in, columna):
  return dataframe_in[columna].apply(lambda x: round(x * 0.06 + 1, 2))

In [12]:
#Muestre sus resultados aquí
for columna in ['math score' , 'reading score' , 'writing score']:
    df_grades[columna] = nota_chilena(df_grades, columna)

df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score  
0              3.76  
3              5.86  
5              4.78  
6              5.56  
7              6.22  
...             ...  
1392           5.26  
1394           3.16  
1396           5.86  
1397           4.72  
1399           5.56  

[1000 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [13]:
#Escriba su código aquí
def alumnos_promedio(dataframe_in):
  asignaturas = ['math score', 'reading score', 'writing score']
  dataframe_in['GPA'] = dataframe_in[asignaturas].mean(axis=1).round(2)
  return dataframe_in

In [14]:
#Muestre sus resultados aquí
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
3  5.78  
5  5.04  
6  5.64  
7  6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [15]:
#Escriba su código aquí
def reprobados_destacados(dataframe_in):
    df_reprobados = dataframe_in[dataframe_in['GPA'] < 4.0]
    df_reprobados_r = dataframe_in.copy()
    df_reprobados_r.loc[df_reprobados_r['GPA'] < 4.0, 'GPA'] = 'R'
    df_destacados = dataframe_in[(dataframe_in['math score'] > 5.8) & (dataframe_in['reading score'] > 5.8) & (dataframe_in['writing score'] > 5.8)]
    df_destacados = df_destacados.sort_values(by='GPA', ascending=False)
    return df_reprobados, df_reprobados_r, df_destacados


In [16]:
# Muestre sus resultados aquí
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

<ipython-input-15-d05611400cd6>:5: FutureWarning:

Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'R' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.



names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
13        Mikki Amaya  female        group A                 high school   
26      Samuel Dougal    male        group C             master's degree   
55    Rodger Talavera    male        group B                 high school   
63     Morris Johnson    male        group A          associate's degree   
...               ...     ...            ...                         ...   
1371   Charles Harian    male        group C            some high school   
1373    Wendy Gunnell  female        group D            some high school   
1389      Diana Costa  female        group C                 high school   
1391      Amy Sieving  female        group C                 high school   
1394    Wanda Russell  female        group B                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
13    free/reduced               completed        3.04           3.88   
26    free/reduced               completed        3.76           3.52   
55    free/reduced                    none        3.16           2.74   
63    free/reduced               completed        3.40           4.30   
...            ...                     ...         ...            ...   
1371  free/reduced               completed        3.70           4.12   
1373  free/reduced                    none        2.62           3.04   
1389  free/reduced                    none        3.46           3.76   
1391      standard                    none        2.74           2.74   
1394  free/reduced               completed        2.38           3.64   

      writing score   GPA  
0              3.76  3.58  
13             3.46  3.46  
26             3.76  3.68  
55             2.62  2.84  
63             4.18  3.96  
...             ...   ...  
1371           3.94  3.92  
1373           2.92  2.86  
1389           3.58  3.60  
1391           2.80  2.76  
1394           3.16  3.06  

[103 rows x 10 columns]

In [17]:
dataframe2

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score   GPA  
0              3.76     R  
3              5.86  5.78  
5              4.78  5.04  
6              5.56  5.64  
7              6.22  6.34  
...             ...   ...  
1392           5.26   5.3  
1394           3.16     R  
1396           5.86  5.42  
1397           4.72  4.54  
1399           5.56  5.48  

[1000 rows x 10 columns]

In [18]:
dataframe3

names  gender race/ethnicity parental level of education  \
1352      Sandra Pompey  female        group E          associate's degree   
384   Elizabeth Beasley  female        group E           bachelor's degree   
677       Richard Pryor    male        group E           bachelor's degree   
475           Emma Gray  female        group E           bachelor's degree   
1276     Angela Cressey  female        group D                some college   
...                 ...     ...            ...                         ...   
949      Rolando Widmer    male        group C                 high school   
296   Lucille Stanphill  female        group C                 high school   
300        Steven Kelly    male        group D                some college   
1307        Peter Perez    male        group D             master's degree   
96          Harry Berry    male        group B          associate's degree   

         lunch test preparation course  math score  reading score  \
1352  standard                    none        7.00           7.00   
384   standard                    none        7.00           7.00   
677   standard               completed        7.00           7.00   
475   standard               completed        6.94           7.00   
1276  standard                    none        6.88           7.00   
...        ...                     ...         ...            ...   
949   standard               completed        5.92           6.04   
296   standard                    none        5.86           6.04   
300   standard                    none        5.86           5.92   
1307  standard                    none        5.86           5.86   
96    standard               completed        5.86           5.92   

      writing score   GPA  
1352           7.00  7.00  
384            7.00  7.00  
677            7.00  7.00  
475            7.00  6.98  
1276           6.94  6.94  
...             ...   ...  
949            5.92  5.96  
296            5.92  5.94  
300            6.04  5.94  
1307           6.04  5.92  
96             5.92  5.90  

[110 rows x 10 columns]

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [19]:
# Escriba su código para quintiles aquí

df_grades['GPA cuantiles'] = pd.qcut(df_grades['GPA'], q=5, labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])


"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""

"\nNOTA: La variable del dataframe que contenga los quintiles debe llamarse\n'GPA cuantiles' para que el gráfico funcione.\n"

In [20]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

```
Se puede observar que existe una tendencia para los hijos que tienen padres con un nivel educacional mayor tienen mejores calificaciones. A medida que los padres tienen un nivel educacional menor, las notas de los hijos tienden a normalizarse.
```

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

**Respuesta:**

In [21]:
# Inserte su respuesta aquí
df_grades_1 = pd.read_json(path + "students_grades_1.json")
df_grades_2 = pd.read_json(path + "students_grades_2.json")
df_grades = pd.concat([df_grades_1, df_grades_2])

df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
998      Richard Young    male        group D                 high school   
999      Wanda Russell  female        group B                 high school   
1000    Marina Zeigler  female        group C           bachelor's degree   
1001     Laurie Carter  female        group B            some high school   
1002      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
1         standard               completed        5.80           5.68   
2         standard                    none        5.38           4.96   
3         standard               completed        5.86           5.50   
4         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
998       standard                    none        5.14           5.50   
999   free/reduced               completed        2.38           3.64   
1000  free/reduced               completed        4.96           5.44   
1001      standard               completed        4.24           4.66   
1002      standard               completed        5.08           5.80   

      writing score   GPA GPA cuantiles  
0              3.76  3.58      Bajísimo  
1              5.86  5.78          Alto  
2              4.78  5.04         Medio  
3              5.56  5.64          Alto  
4              6.22  6.34      Muy alto  
...             ...   ...           ...  
998            5.26  5.30         Medio  
999            3.16  3.06      Bajísimo  
1000           5.86  5.42          Alto  
1001           4.72  4.54          Bajo  
1002           5.56  5.48          Alto  

[1000 rows x 11 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos]**


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


**Respuesta:**

In [22]:
#Calcular el promedio de las notas y guardarlo en una variable GPA (grade point average).
df_grades['GPA'] = df_grades[['math score', 'reading score', 'writing score']].mean(axis=1)

#Hacer una simplificación a través de un mapeo (investigar el método map()) de la variable parental level of education según la siguiente conversión.
#Los resultados de este mapeo deben ser guardados en la columna simple parental level of education.
mapeo = {
    'some high school': 'school',
    'some college': 'school',
    'high school': 'school',
    "bachelor's degree": 'college',
    "associate's degree": 'college',
    "master's degree": 'postgraduate'
}
df_grades['simple parental level of education'] = df_grades['parental level of education'].map(mapeo)

#Agregar según 2 niveles: race/ethnicity y simple parental level of education para obtener el promedio de las notas.
#Agregar según 2 niveles: race/ethnicity y simple parental level of education para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
resumen = df_grades.groupby(['race/ethnicity', 'simple parental level of education']).agg(
    math_score=('math score', 'mean'),
    reading_score=('reading score', 'mean'),
    writing_score=('writing score', 'mean'),
    GPA=('GPA', 'mean'),
    count=('GPA', 'count')
  ).reset_index()

#Se redondean los valores que exceden los dos decimales
resumen['GPA'] = resumen['GPA'].round(2)
resumen['math_score'] = resumen['math_score'].round(2)
resumen['reading_score'] = resumen['reading_score'].round(2)
resumen['writing_score'] = resumen['writing_score'].round(2)


#Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.
total_alumnos = len(df_grades)
resumen['percentage'] = (resumen['count'] / total_alumnos * 100).round(2).astype(str) + '%'

resumen

race/ethnicity simple parental level of education  math_score  \
0         group A                            college        4.83   
1         group A                       postgraduate        4.46   
2         group A                             school        4.65   
3         group B                            college        5.03   
4         group B                       postgraduate        5.03   
5         group B                             school        4.69   
6         group C                            college        5.03   
7         group C                       postgraduate        5.02   
8         group C                             school        4.74   
9         group D                            college        5.06   
10        group D                       postgraduate        5.35   
11        group D                             school        4.99   
12        group E                            college        5.53   
13        group E                       postgraduate        5.48   
14        group E                             school        5.35   

    reading_score  writing_score   GPA  count percentage  
0            5.05           4.95  4.94     26       2.6%  
1            4.88           5.06  4.80      3       0.3%  
2            4.81           4.66  4.71     60       6.0%  
3            5.24           5.16  5.14     61       6.1%  
4            5.81           5.63  5.49      6       0.6%  
5            4.90           4.79  4.79    123      12.3%  
6            5.36           5.33  5.24    118      11.8%  
7            5.23           5.17  5.14     19       1.9%  
8            5.00           4.89  4.88    182      18.2%  
9            5.22           5.24  5.17     78       7.8%  
10           5.63           5.78  5.59     23       2.3%  
11           5.13           5.11  5.08    161      16.1%  
12           5.45           5.43  5.47     57       5.7%  
13           5.93           5.83  5.74      8       0.8%  
14           5.27           5.11  5.25     75       7.5%

In [23]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

    Como se puede ver en los gráficos, se puede evidenciar que para el caso de los padres con un mayor nivel educacional simple, los hijos también tienen notas más altas. Con respecto a la etnicidad, se puede observar que las personas pertenecientes al grupo E poseen una media un poco más elevada en general que los otros grupos, excepto en el caso de padres con un nivel de "postgraduate" donde esta media posee una gran diferencia con respecto a las otras, en otros casos los niveles andan todos en un valor similar. Por otro lado las personas correspondientes al grupo A son las que tienen una media menor, donde se tiene el mismo comportamiento antes mencionado.

**Resultados Esperados**

Esta tabla ilustra la estructura del dataframe a conseguir. Los valores son referenciales, no necesariamente son los esperados.

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.83 |            5.05 |            4.95 |  4.94 |      26 | 2.6 %        |
|  1 |                  | postgraduate                         |         4.46 |            4.88 |            5.06 |  4.80 |       3 | 0.3 %        |
|  2 |                  | school                               |         4.65 |            4.81 |            4.66 |  4.71 |      60 | 6.0 %        |
|  3 | group B          | college                              |         5.03 |            5.24 |            5.16 |  5.14 |      61 | 6.1 %        |
|  4 |                  | postgraduate                         |         5.03 |            5.81 |            5.63 |  5.49 |       6 | 0.6 %        |
|  5 |                  | school                               |         4.69 |            4.90 |            4.79 |  4.79 |     123 | 12.3 %       |
|  6 | group C          | college                              |         5.03 |            5.36 |            5.33 |  5.24 |     118 | 11.8 %       |
|  7 |                  | postgraduate                         |         5.02 |            5.23 |            5.17 |  5.14 |      19 | 1.9 %        |
|  8 |                  | school                               |         4.74 |            5.00 |            4.89 |  4.88 |     182 | 18.2 %       |
|  9 | group D          | college                              |         5.06 |            5.22 |            5.24 |  5.17 |      78 | 7.8 %        |
| 10 |                  | postgraduate                         |         5.35 |            5.63 |            5.78 |  5.59 |      23 | 2.3 %        |
| 11 |                  | school                               |         4.99 |            5.13 |            5.11 |  5.08 |     161 | 16.1 %       |
| 12 | group E          | college                              |         5.53 |            5.45 |            5.43 |  5.47 |      57 | 5.7 %        |
| 13 |                  | postgraduate                         |         5.48 |            5.93 |            5.83 |  5.74 |       8 | 0.8 %        |
| 14 |                  | school                               |         5.35 |            5.27 |            5.11 |  5.25 |      75 | 7.5 %        |


### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene registros de alumnos que ya no forman parte de la universidad, por lo que no deben ser parte de este análisis.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados o **alumnos que no pertenecen a la universidad**).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

**Respuesta:**

In [24]:
# Inserte su respuesta aquí
other_grades = pd.read_csv(path + "other_grades.csv")

df_grades_outer = pd.merge(df_grades, other_grades, on='names', how='outer')
df_grades_left = pd.merge(df_grades, other_grades, on='names', how='left')
df_grades_right = pd.merge(df_grades, other_grades, on='names', how='right')
df_grades_inner = pd.merge(df_grades, other_grades, on='names', how='inner')

other_grades

names  science score  history score
0           Pam Otoole           4.42           4.00
1          Diane Olson           3.88           4.80
2       Sarah Flannery           5.32           6.10
3     Stanley Chappell           3.82           3.70
4        Arnold Dodson           4.60           5.20
...                ...            ...            ...
1145     Carson Brooks           4.72           1.54
1146         Ruby Reed           1.71           4.59
1147    Zachary Torres           4.48           6.92
1148       Eliana Gray           6.91           5.41
1149       Gavin James           6.60           6.06

[1150 rows x 3 columns]

In [25]:
df_grades_outer

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1145     Carson Brooks     NaN            NaN                         NaN   
1146         Ruby Reed     NaN            NaN                         NaN   
1147    Zachary Torres     NaN            NaN                         NaN   
1148       Eliana Gray     NaN            NaN                         NaN   
1149       Gavin James     NaN            NaN                         NaN   

         lunch test preparation course  math score  reading score  \
0     standard                    none        3.22           3.76   
1     standard               completed        5.80           5.68   
2     standard                    none        5.38           4.96   
3     standard               completed        5.86           5.50   
4     standard               completed        6.64           6.16   
...        ...                     ...         ...            ...   
1145       NaN                     NaN         NaN            NaN   
1146       NaN                     NaN         NaN            NaN   
1147       NaN                     NaN         NaN            NaN   
1148       NaN                     NaN         NaN            NaN   
1149       NaN                     NaN         NaN            NaN   

      writing score   GPA GPA cuantiles simple parental level of education  \
0              3.76  3.58      Bajísimo                             school   
1              5.86  5.78          Alto                            college   
2              4.78  5.04         Medio                             school   
3              5.56  5.64          Alto                             school   
4              6.22  6.34      Muy alto                             school   
...             ...   ...           ...                                ...   
1145            NaN   NaN           NaN                                NaN   
1146            NaN   NaN           NaN                                NaN   
1147            NaN   NaN           NaN                                NaN   
1148            NaN   NaN           NaN                                NaN   
1149            NaN   NaN           NaN                                NaN   

      science score  history score  
0              2.50           3.80  
1              4.36           5.20  
2              4.48           3.80  
3              5.32           5.00  
4              6.64           4.60  
...             ...            ...  
1145           4.72           1.54  
1146           1.71           4.59  
1147           4.48           6.92  
1148           6.91           5.41  
1149           6.60           6.06  

[1150 rows x 14 columns]

In [26]:
df_grades_left

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
995      standard                    none        5.14           5.50   
996  free/reduced               completed        2.38           3.64   
997  free/reduced               completed        4.96           5.44   
998      standard               completed        4.24           4.66   
999      standard               completed        5.08           5.80   

     writing score   GPA GPA cuantiles simple parental level of education  \
0             3.76  3.58      Bajísimo                             school   
1             5.86  5.78          Alto                            college   
2             4.78  5.04         Medio                             school   
3             5.56  5.64          Alto                             school   
4             6.22  6.34      Muy alto                             school   
..             ...   ...           ...                                ...   
995           5.26  5.30         Medio                             school   
996           3.16  3.06      Bajísimo                             school   
997           5.86  5.42          Alto                            college   
998           4.72  4.54          Bajo                             school   
999           5.56  5.48          Alto                             school   

     science score  history score  
0             2.50            3.8  
1             4.36            5.2  
2             4.48            3.8  
3             5.32            5.0  
4             6.64            4.6  
..             ...            ...  
995           4.72            4.2  
996           1.96            3.6  
997           4.54            4.1  
998           3.58            4.3  
999           4.66            5.3  

[1000 rows x 14 columns]

In [27]:
df_grades_right

names  gender race/ethnicity parental level of education  \
0           Pam Otoole  female        group B           bachelor's degree   
1          Diane Olson  female        group C                some college   
2       Sarah Flannery  female        group B             master's degree   
3     Stanley Chappell    male        group A          associate's degree   
4        Arnold Dodson    male        group C                some college   
...                ...     ...            ...                         ...   
1145     Carson Brooks     NaN            NaN                         NaN   
1146         Ruby Reed     NaN            NaN                         NaN   
1147    Zachary Torres     NaN            NaN                         NaN   
1148       Eliana Gray     NaN            NaN                         NaN   
1149       Gavin James     NaN            NaN                         NaN   

             lunch test preparation course  math score  reading score  \
0         standard                    none        5.32           5.32   
1         standard               completed        5.14           6.40   
2         standard                    none        6.40           6.70   
3     free/reduced                    none        3.82           4.42   
4         standard                    none        5.56           5.68   
...            ...                     ...         ...            ...   
1145           NaN                     NaN         NaN            NaN   
1146           NaN                     NaN         NaN            NaN   
1147           NaN                     NaN         NaN            NaN   
1148           NaN                     NaN         NaN            NaN   
1149           NaN                     NaN         NaN            NaN   

      writing score   GPA GPA cuantiles simple parental level of education  \
0              5.44  5.36          Alto                            college   
1              6.28  5.94      Muy alto                             school   
2              6.58  6.56      Muy alto                       postgraduate   
3              3.64  3.96      Bajísimo                            college   
4              5.50  5.58          Alto                             school   
...             ...   ...           ...                                ...   
1145            NaN   NaN           NaN                                NaN   
1146            NaN   NaN           NaN                                NaN   
1147            NaN   NaN           NaN                                NaN   
1148            NaN   NaN           NaN                                NaN   
1149            NaN   NaN           NaN                                NaN   

      science score  history score  
0              4.42           4.00  
1              3.88           4.80  
2              5.32           6.10  
3              3.82           3.70  
4              4.60           5.20  
...             ...            ...  
1145           4.72           1.54  
1146           1.71           4.59  
1147           4.48           6.92  
1148           6.91           5.41  
1149           6.60           6.06  

[1150 rows x 14 columns]

In [28]:
df_grades_inner

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
995      standard                    none        5.14           5.50   
996  free/reduced               completed        2.38           3.64   
997  free/reduced               completed        4.96           5.44   
998      standard               completed        4.24           4.66   
999      standard               completed        5.08           5.80   

     writing score   GPA GPA cuantiles simple parental level of education  \
0             3.76  3.58      Bajísimo                             school   
1             5.86  5.78          Alto                            college   
2             4.78  5.04         Medio                             school   
3             5.56  5.64          Alto                             school   
4             6.22  6.34      Muy alto                             school   
..             ...   ...           ...                                ...   
995           5.26  5.30         Medio                             school   
996           3.16  3.06      Bajísimo                             school   
997           5.86  5.42          Alto                            college   
998           4.72  4.54          Bajo                             school   
999           5.56  5.48          Alto                             school   

     science score  history score  
0             2.50            3.8  
1             4.36            5.2  
2             4.48            3.8  
3             5.32            5.0  
4             6.64            4.6  
..             ...            ...  
995           4.72            4.2  
996           1.96            3.6  
997           4.54            4.1  
998           3.58            4.3  
999           4.66            5.3  

[1000 rows x 14 columns]

    El mejor método para este caso dependería del objetivo, ya que si nos interesa más el tener los mismos alumnos independiente de si tienen la nota nueva o no se recomendaría usar el left join.
    Por otro lado si nos interesa más el caso de que todos los alumnos que presentan nota en las evaluaciones de historia y ciencias se encuentren en el dataframe final y si no tienen estas notas que no aparezcan, el mejor método a elegir correspondería a un inner join.
    Por como está explicado el problema sería mejor usar el inner join, ya que se da a entender que los alumnos presentes en el dataframe original si se encuentran en el nuevo dataframe por lo que no habría un problema de eliminación de estudiantes al usar este método.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>